src/Approximations/overapproximate.jl

```diff
@@ -516,3 +516,67 @@ function overapproximate(cap::Intersection{N,
     return overapproximate(swap(cap), dir; kwargs...)
 end
 
+"""
+    overapproximate(lm::LinearMap{N, <:CartesianProductArray},
+                      ::Type{<:CartesianProductArray}, oa=Hyperrectangle)  where {N}
+
+Decompose a lazy linear map of a cartesian product array.
+
+### Input
+
+- `lm`   -- lazy linear map of cartesian product array
+- `CartesianProductArray` -- type for dispatch
+- `oa` -- approximation option for decomposition
+
+### Output
+
+A `CartesianProductArray` representing the decomposed linear map.
+"""
+function overapproximate(lm::LinearMap{N, <:CartesianProductArray},
+                           ::Type{<:CartesianProductArray}, oa=Hyperrectangle) where {N}
+
+    cpa = lm.X
+    M = lm.M
+
+    array = allocate_result(oa, N)
+    sizehint!(array, length(cpa.array))
+
+    row_start_ind, row_end_ind = 1, 0
+    for bi in cpa.array
+        row_end_ind += dim(bi)
+        ms = blocks_linear_map(cpa, M, row_start_ind : row_end_ind)
+        push!(array, overapproximate(ms, oa))
+        row_start_ind += row_end_ind
+    end
+
+    result = CartesianProductArray(array)
+    return result
+end
+
+
+function blocks_linear_map(cpa::CartesianProductArray,
+                           M::AbstractMatrix{N},
+                           row_range::UnitRange{Int}) where {N}
+    col_start_ind, col_end_ind = 1, 0
+    array = Vector{LazySet{N}}()
```

Comment:
Since we always put a `LinearMap` in this array, we can change the type to `Vector{LazySet{N}}`

Reply:
I would try to use the type information of `cpa` as well.

Reply:
In general, each block of `cpa` might have different type.

Reply:
Actually why don't you make `array` a `MinkowskiSumArray`?

Reply:
See https://github.com/JuliaReach/LazySets.jl/pull/1419#discussion_r290609335

Reply:
It *is* possible to create a CPA with different set types, but i think it is not the correct approach -- with performance in mind -- to write the `overapproximate` functions in a type-agnostic way straight from the beginning. Moreover, in `Reachability.jl` benchmarks we have not used different set types for different blocks that i remember.

Reply:
For instance in `Reachability.jl`, i think we should move in the direction of writing type-stable versions of the decomposed algo and the hybrid scheme.

Reply:
I think the default decomposition using the block structure `[2,  1]` will choose the types `[Hyperrectangle, Interval]`, which is mixed.

Reply:
Let me reformulate: there are cases where the set type is mixed (the `cpa` is of type `CartesianProductArray{N, LazySet{N}}`, and others where it isn't (the `cpa` is of type `CartesianProductArray{N, Hyperrectangle{N}}` or other concrete type).

This implementation covers the first case, but ignores the information of the second one when there is some.

Reply:
So if the input is `CartesianProductArray{N, T}`, then we should create a `Vector{LinearMap{N, T}}`?

Reply:
Yes,

```julia
function blocks_linear_map(cpa::CartesianProductArray{N, T},
                           M::AbstractMatrix{N},
                           row_range::UnitRange{Int}) where {N, T<:LazySet{N}}
    col_start_ind, col_end_ind = 1, 0
    array = Vector{LinearMap{N, T}}()
...
```

Reply:
Below is the full code. This implementation receives a typed output set type, such as the cartesian product array of intervals, or hyperrectangles, or `AbstractHyperrectangle`. There is a second signature that allows to receive a third argument, the template direction. 


```julia
function overapproximate(lm::LinearMap{N, CartesianProductArray{N, T}},
                           ::Type{CartesianProductArray{N, O}}) where {N, T<:LazySet{N}, O<:LazySet{N}}

    M, cpa = lm.M, lm.X
    array = Vector{O}(undef, length(cpa.array))
    return _overapproximate_lm_cpa!(M, cpa, array, O)
end

function overapproximate(lm::LinearMap{N, CartesianProductArray{N, T}},
                           ::Type{CartesianProductArray{N, HPolytope{N}}},
                        dir::AbstractDirections{N}) where {N, T<:LazySet{N}}
    M, cpa = lm.M, lm.X
    array = Vector{HPolytope{N}}(undef, length(cpa.array))
    return _overapproximate_lm_cpa!(M, cpa, array, O)
end

function _overapproximate_lm_cpa!(M, cpa, array, O)
    row_start_ind, row_end_ind = 1, 0
    @inbounds for (i, bi) in enumerate(cpa.array)
        row_end_ind += dim(bi)
        ms = blocks_linear_map(cpa, M, row_start_ind : row_end_ind)
        array[i] = overapproximate(ms, O)
        row_start_ind += row_end_ind
    end
    return CartesianProductArray(array)
end

function blocks_linear_map(cpa::CartesianProductArray{N, T},
                           M::AbstractMatrix{N},
                           row_range::UnitRange{Int}) where {N, T<:LazySet{N}}
    array = Vector{LinearMap{N, <:T}}(undef, length(cpa.array))

    col_start_ind, col_end_ind = 1, 0
    @inbounds for (i, bi) in enumerate(cpa.array)
        col_end_ind += dim(bi)
        array[i] = LinearMap(M[row_range, col_start_ind : col_end_ind], bi)
        col_start_ind += col_end_ind
    end

    return MinkowskiSumArray(array)
end
```

For the tests this will require to write for example,

```julia
d_oa_d_hp = overapproximate(lm, CartesianProductArray{N, Hyperrectangle{N}})
```

A "lazy" signature  `d_oa_d_hp = overapproximate(lm, CartesianProductArray)` which decomposes into the CPA of hyperrectangles by default is (maybe) doable without too much changes, perhaps with `invoke`, but i don't know if we should bother.

Reply:
I believe this code doesn't cover mixed set type.

Reply:
> I believe this code doesn't cover mixed set type.

Why? If you are worried about the type `T`: that is (by definition) a common supertype of all elements in `cpa.array`.

Reply:
> I believe this code doesn't cover mixed set type.

The best way is to give a counterexample in the form of a test case.

Reply:
We actually have a test with mixed set types.